airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/source.py

```diff
@@ -233,6 +232,46 @@ def stream_slices(
                 start_date += datetime.timedelta(days=self.config["window_in_days"])
 
 
+class PivotReport(GoogleAnalyticsDataApiBaseStream):
+    def request_body_json(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Mapping]:
+        payload = super().request_body_json(stream_state, stream_slice, next_page_token)
+        payload["pivots"] = self.config["pivots"]
+        return payload
+
+    def path(
+        self, *, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return f"properties/{self.config['property_id']}:runPivotReport"
+
+
+class CohortReportMixin:
+    cursor_field = []
+
+    def stream_slices(
+        self, *, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        yield from [None]
+
+    def request_body_json(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Mapping]:
+        # https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/CohortSpec#Cohort.FIELDS.date_range
+        # In a cohort request, this dateRange is required and the dateRanges in the RunReportRequest or RunPivotReportRequest
```

Comment:
Why you delete dateRange from payload if you said that is required for cohort request. And this mixin is parent only for cohort report.

Reply:
These are different date ranges. The one in the root of the payload should be unspecified because there is another one inside the "cohortSpec" object (see schema for details). 

This mixin is one of the base classes for the dynamically created report if there's a "cohortSpec" object in the report definition. It can be either a regular report or a pivot report.